sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp

```diff
@@ -0,0 +1,365 @@
+    };
+
+    static azure::core::http::Request StageBlockListConstructRequest(
+        /*const*/ std::string& url,
+        /*const*/ StageBlockListOptions& options)
+    {
+      azure::core::http::Request request(azure::core::http::HttpMethod::Put, url);
+      request.addQueryParameter(k_QUERY_COMP, k_QUERY_BLOCK_LIST);
+      ApplyBasicHeaders(options, request);
+      //TODO: avoid the data copy here.
+      block_list_writer writer;
+      std::string body = writer.write(options.BlockList);
+      uint8_t* buffer = new uint8_t[body.size() + 1];
+      memcpy_s(buffer, body.size() + 1, body.c_str(), body.size() + 1);
+      azure::core::http::BodyBuffer* bodyBuffer = new azure::core::http::BodyBuffer(buffer, body.size());
+      request.setBodyBuffer(bodyBuffer);
+      return request;
+    }
+
+
+
+    // TODO: should return azure::core::http::Response<BlockInfo> instead
+    static BlockInfo StageBlockListParseResponse(/*const*/ azure::core::http::Response& response)
+    {
+      BlockInfo info;
+      info.ContentMd5 = GetHeaderValue(response.getHeaders(), k_HEADER_CONTENT_MD5);
+      info.ContentCrc64 = GetHeaderValue(response.getHeaders(), k_HEADER_X_MS_CONTENT_CRC64);
+      ParseBasicResponseHeaders(response.getHeaders(), info);
+      return info;
+    }
+
+    // TODO: should return azure::core::http::Response<BlockInfo> instead
+    static BlockInfo StageBlockList(
+        // TODO: Context and Pipeline should go here
+        /*const*/ std::string& url,
+        /*const*/ StageBlockListOptions& options)
+    {
+        //TODO Manage the memory that is allocated when constructing the request.
+      return StageBlockListParseResponse(
+          azure::core::http::Client::send(StageBlockListConstructRequest(url, options)));
+    }
+
+  private:
+    class block_list_writer : public Azure::Storage::Common::XML::xml_writer
+    {
+    public:
+      block_list_writer() {}
+
+      std::string write(const std::vector<std::pair<BlockType, std::string>>& blockList)
+      {
+        std::ostringstream outstream;
+        initialize(outstream);
+
+        write_start_element(k_XML_TAG_BLOCK_LIST);
+
+        for (auto block = blockList.cbegin(); block != blockList.cend(); ++block)
+        {
+          std::string tag;
+          switch (block->first)
+          {
+            case BlockType::Committed:
+              tag = k_XML_TAG_COMMITTED;
+              break;
+
+            case BlockType::Uncommitted:
+              tag = k_XML_TAG_UNCOMMITTED;
+              break;
+
+            case BlockType::Latest:
+              tag = k_XML_TAG_LATEST;
+              break;
+          }
+
+          write_element(tag, block->second);
+        }
+
+        finalize();
+        return outstream.str();
+      }
+    };
+  };
+
+private:
+  static void AddMetadata(
+      /*const*/ std::map<std::string, std::string>& metadata,
+      azure::core::http::Request& request)
+  {
+    for (auto pair : metadata)
```

Comment:
You may want `const auto&` to avoid copy.

Reply:
Agreed, this is a lazy mistake :(